### ISSUE TITLE
Toy "Demo" Illustrating the Flow

### ISSUE THREAD

**Issue by 0x4007** [ubiquity/ubq.fi#29]
![dxos org_](https://github.com/ubiquity/ubq.fi/assets/4975670/3d7044f1-ee8f-45c1-a933-53028fa904bc)

[DXOS](https://dxos.org/) demonstrates their tech above-the-fold, which is the most concise way to explain their offering.

We should make a simple example like this that is a simplified model of our flow. 

For example:

- There is a simple description "Move the box" 
- There already is a priority level set
- User is prompted to set a time estimate
- It shows the price label
- It prompts the user to "start" the task
- It renders a box that you can drag and drop to a goal box. 
- As soon as its complete, a payment permit is shown with the reward
- The user can actually claim an NFT etc. 

This can be handled in a separate repo and we can iframe it in since this seems a bit involved. Also I think we need to create an NFT contract for this demo.

**Comment by rndquu** [comment]
We could start simple and use smth like https://tourguidejs.com/ (you may click on the "Start tour" button to see how it works).

So we could:
1. Create a separate page like `demo.ubq.fi`
2. Paste plain html of any issue (like [this](https://github.com/ubiquity/pay.ubq.fi/issues/247) one) with generated permit so that demo page looked like a github page
3. Automatically start an onboarding tour (via https://tourguidejs.com) when the page is loaded

**Comment by 0x4007** [comment]
This tour guide can be an extension of this vision but the original idea I had in mind was to try and explain the concept alone. 

Yes it is nice to see the concept within the context of GitHub so that it's more tangible, but if we took just the essence (do work, bot eval[^1^], get paid) it would be more in line with what I am proposing. 

I think both ideas can work together though and it would be really nice to do tour guide as well. 

[^1^]: perhaps to explain qualitative evaluation (like conversation rewards) we could observe how much of a straight path they took to move the box. Or quantitative score could be the user's time to completion etc. it could be interesting to personify the UbiquiBot and show that it's judging the work.

**Comment by rndquu** [comment]
Also a good example of quick start: https://www.buildbear.io/resources/quick-guide

**Comment by rndquu** [comment]
@pbkompasz Could you elaborate on how exactly you're going to solve this issue?

**Comment by pbkompasz** [comment]
@rndquu 
I think the best option would be breaking down the quick start guide into steps, similarty to buildbear.io 
1. First step is a list of demo tasks are shown, from which the user can select one
2. After selecting one the task page is rendered
3. User can interact with the task: set time estimate, prompt help, start the task, etc. The bot responses are rendered after each prompt
4. After the job is "completed" and the permit to redeem the NFT is shown
5. The user can complete the demo or claim the NFT

Where it is possible it would show a tour of the features using tourguidejs.

**Comment by 0x4007** [comment]
I think that's more complicated than it needs to be. Perhaps stick with the box example in the spec.

**Comment by nick-dev123** [comment]
@0x4007 I am going to create a separate [repo](https://github.com/nick-dev123/ubq-fi-demo) for the demo.
Will the bot keep track of progress in that?
For the implementation I’m working on, the layout consists of two columns with a center section. Users can select a job from the available list in the left column, which then moves to the center for interaction (where users can send prompts). Once completed, they can drag the job to the right column to claim the NFT.  Does that sound like what you were thinking, or did you have a different idea in mind?

**Comment by 0x4007** [comment]
> @0x4007 I am going to create a separate [repo](https://github.com/nick-dev123/ubq-fi-demo) for the demo.
> 
> Will the bot keep track of progress in that?
> 

Don't understand your question

> For the implementation I’m working on, the layout consists of two columns with a center section. Users can select a job from the available list in the left column, which then moves to the center for interaction (where users can send prompts). 

What are prompts

> Once completed, they can drag the job to the right column to claim the NFT.  Does that sound like what you were thinking, or did you have a different idea in mind?

Seems close but need clarity

**Comment by nick-dev123** [comment]
> > @0x4007 I am going to create a separate [repo](https://github.com/nick-dev123/ubq-fi-demo) for the demo.
> > Will the bot keep track of progress in that?
> 
> Don't understand your question
> 
I meant that will it un-assign me due to no updates from a linked pull request?

> > For the implementation I’m working on, the layout consists of two columns with a center section. Users can select a job from the available list in the left column, which then moves to the center for interaction (where users can send prompts).
> 
> What are prompts
> 

Initially `/help`, `/start`, `/wallet` and `/time` to set the time estimate.

> > Once completed, they can drag the job to the right column to claim the NFT.  Does that sound like what you were thinking, or did you have a different idea in mind?
> 
> Seems close but need clarity

I will let you know when I have an MPV ready.

**Comment by 0x4007** [comment]
Not sure about /time might be misleading cause we don't have that feature. Everything else seems good. We will make sure you are assigned if you continue to work on it

**Comment by rndquu** [comment]
@nick-dev123 You better use https://github.com/ubiquity/ts-template as a repo template which already has cloudflare deployment previews. Later we can move your repo to the https://github.com/ubiquity organization.

**Comment by nick-dev123** [comment]
@0x4007 I implemented the main functionalities.
The styling is basic and it's just to showcase the functionalities, I was thinking the final design on the left should match the [Devpool Directory](https://devpool.directory/) list and on the right it would smaller cards with the job name and a button to redeem the reward.
This is the recording of a sample job where the user has to set their wallet address and start the job, after which it can drag it to the completed jobs list.
https://github.com/user-attachments/assets/1a0e0c87-425f-4e1e-b6cb-23095d1d2065

**Comment by 0x4007** [comment]
UI part seems pretty great! Good work so far. @rndquu any advice on the NFT minting part? 

We can ask the user to "register" their wallet address manually and the front end code can generate the NFT "mint permit"? 

It can all be fake and not secure so that it's simple, but the point is that the user should be able to claim/mint a real NFT reward for it to make sense to newcomers. 

For example we could fake it by just checking in the UI if the wallet address matches. And if it does, to call a generic mint function that would still work for any address if you directly called the contract etc. so that way we don't need to implement the real NFT mint permit (unless that's already completed? I vaguely recall it is but was never actually integrated into our system)

**Comment by nick-dev123** [comment]
@0x4007 Should I match the style of the list from the Devpool exactly e.g. font size, colors, hover effect, etc.?

**Comment by 0x4007** [comment]
Would be ideal if you can lift the same CSS, yes.

**Comment by rndquu** [comment]
> @rndquu any advice on the NFT minting part?

@0x4007 Should NFT minting (i.e. permits generation) be part of the demo page?

@nick-dev123 If NFT minting should be present in the demo page then you may:
1. Install https://www.npmjs.com/package/@ubiquibot/permit-generation
2. Call [generatePayoutPermit](https://github.com/ubiquity-os/permit-generation/blob/0b3355d415341e57aceac568ac82e752a10a1328/src/handlers/generate-payout-permit.ts#L13) with type `ERC721` to generate an NFT permit object. You should place all dummy values except for the NFT receiver address (which you'd get via the `/wallet 0x...` command in the UI).
3. [Convert](https://github.com/ubiquity-os-marketplace/conversation-rewards/blob/139d2a2a87e4db9d007fa8f8876f80aa8645f71b/src/parser/permit-generation-module.ts#L121) NFT permit object to URL and display in the UI

**Comment by 0x4007** [comment]
> @0x4007 Should NFT minting (i.e. permits generation) be part of the demo page?

Yes it's part of the spec. I figured use the nft solution we have instead of making a custom one because I assume it would be faster to set up.

**Comment by nick-dev123** [comment]
@0x4007 The permit URL should be shown as a notification right? Or as a response in the "chat"?
Also how will the demo be integrated into the existing project, will it be on a separate page or iframed into an existing one?

**Comment by 0x4007** [comment]
> @0x4007 The permit URL should be shown as a notification right? Or as a response in the "chat"?
> 

The closer that we can simulate it to the real experience, the better. If you can make it a response in the chat that's probably best. 

> Also how will the demo be integrated into the existing project, will it be on a separate page or iframed into an existing one?

Depends on complexity. I can probably lift the code and get it working eventually but first step is just to iframe it in.

**Comment by nick-dev123** [comment]
@0x4007 What about on mobile?
I think the three columns is too much to show at once, I was thinking of showing first the select jobs list, then when a user selects a job it slides down to show the chat interface, and finally to the completed jobs list.

**Comment by 0x4007** [comment]
We can consider either omitting some aspects or even linking out for them to experience it on a dedicated page? It's hard to say because I didn't build it but I think it may be reasonable to expect the first iteration to work on desktop only.

**Comment by nick-dev123** [comment]
.

**Comment by nick-dev123** [comment]
@0x4007 Can I get reassigned? I just started implementing the permit generation workflow.

**Comment by 0x4007** [comment]
No. You had about a month to start. You're supposed to start as soon as you write /start.

**Comment by nick-dev123** [comment]
I have started as soon as I got assigned in this [repo](https://github.com/nick-dev123/ubq-fi-demo).
It has almost everything done except the permit generation, but I already started it I just didn't commit and push it yet.
The following is completed:
- generate jobs by creating a `job1.json` file
- jobs can be selected, interacted with and completed
- the styling is almost complete and matches the one in the work.ubq.fi

**Comment by 0x4007** [comment]
Open a pull against your own repo and link it here. Afterwards you can be assigned. You'll be unassigned after seven days of inactivity.

**Comment by 0x4007** [comment]
https://v0.dev can probably generate this for you.

**Comment by luisantoniocrag** [comment]
Hey @0x4007 is this task still available to work on?

**Comment by henalolp** [comment]
@0x4007 I wasn't able to assign myself for this task; I think there might be some issue with the bot.
Can you fix this issue?

**Comment by rndquu** [comment]
@nick-dev123 I've tried how https://github.com/nick-dev123/ubq-fi-demo works, it seems it's missing permit generation. If there exists another repository solving current github issue pls open a PR.

**Comment by nick-dev123** [comment]
@rndquu That's the repo with the demo, it is missing a couple of features and refinements, I had to stop working on it due to being busy. Can I get reassigned and finish it up?

**Comment by rndquu** [comment]
> @rndquu That's the repo with the demo, it is missing a couple of features and refinements, I had to stop working on it due to being busy. Can I get reassigned and finish it up?

You're free to run the `/start` since nobody is assigned to the current task

**Comment by pbkompasz** [comment]
@rndquu Doesn't work. Can you assign me manually?

**Comment by rndquu** [comment]
> @rndquu Doesn't work. Can you assign me manually?

Try to open a PR.

@gentlementlegen Should the `/start` work or contributors must open a PR first?

**Comment by gentlementlegen** [comment]
@rndquu start should work. The kernel on the production branch has been skipping a lot of event forwarding somehow for the last two days, and I cannot find the cause, because it is the same code as the development version which seems to have no issue, sorry about that.
